File: ios/auto-mobile-sdk/Sources/AutoMobileSDK/Network/RetryPolicy.swift
import Foundation

/// Configures exponential backoff with jitter for retrying failed event delivery.
public struct RetryPolicy: Sendable {
    public let maxRetries: Int
    public let baseDelayMs: Int
    public let maxDelayMs: Int

    public init(maxRetries: Int = 3, baseDelayMs: Int = 1000, maxDelayMs: Int = 30_000) {
        self.maxRetries = maxRetries
        self.baseDelayMs = baseDelayMs
        self.maxDelayMs = maxDelayMs
    }

    /// Returns whether to retry and the delay in milliseconds, given a status code and attempt number.
    /// Retries on: network failure (0), timeout (408), rate limiting (429), server errors (500-599).
    public func shouldRetry(statusCode: Int, attempt: Int) -> (shouldRetry: Bool, delayMs: Int) {
        guard attempt < maxRetries else { return (false, 0) }
        let retryable = statusCode == 0 || statusCode == 408 || statusCode == 429 ||
                        (statusCode >= 500 && statusCode <= 599)
        guard retryable else { return (false, 0) }
        let delay = min(baseDelayMs * (1 << attempt), maxDelayMs)
        let jitter = Int.random(in: 0...max(1, delay / 4))
        return (true, delay + jitter)
    }
}

File: android/auto-mobile-sdk/src/main/kotlin/dev/jasonpearson/automobile/sdk/events/RetryPolicy.kt
package dev.jasonpearson.automobile.sdk.events

/**
 * Configures exponential backoff with jitter for retrying failed event delivery.
 *
 * @param maxRetries Maximum number of retry attempts (default 3)
 * @param baseDelayMs Base delay in milliseconds before first retry (default 100)
 * @param maxDelayMs Upper bound on computed delay (default 2000)
 */
internal data class RetryPolicy(
    val maxRetries: Int = 3,
    val baseDelayMs: Long = 100,
    val maxDelayMs: Long = 2000,
) {
  /**
   * Compute the delay for a given retry attempt (0-indexed).
   *
   * Uses exponential backoff (`baseDelayMs * 2^attempt`) clamped to [maxDelayMs], plus random
   * jitter of up to 25% of the clamped delay.
   */
  fun delayForAttempt(attempt: Int): Long {
    val clamped = attempt.coerceIn(0, 30)
    val delay = (baseDelayMs * (1L shl clamped)).coerceAtMost(maxDelayMs)
    val jitter = if (delay > 0) (0..delay / 4).random() else 0L
    return delay + jitter
  }
}
